powertools-parameters/src/main/java/software/amazon/lambda/powertools/parameters/AppConfigProvider.java

```diff
@@ -0,0 +1,228 @@
+        private final String lastConfigurationValue;
+
+        private EstablishedSession(String nextSessionToken, String value) {
+            this.nextSessionToken = nextSessionToken;
+            this.lastConfigurationValue = value;
+        }
+    }
+
+    private final AppConfigDataClient client;
+
+    private final String application;
+
+    private final String environment;
+
+    private final HashMap<String, EstablishedSession> establishedSessions = new HashMap<>();
+
+    AppConfigProvider(CacheManager cacheManager, AppConfigDataClient client, String environment, String application) {
+        super(cacheManager);
+        this.client = client;
+        this.application = application;
+        this.environment = environment;
+    }
+
+
+    /**
+     * Retrieve the parameter value from the AppConfig parameter store.<br />
+     *
+     * @param key key of the parameter. This ties back to AppConfig's 'profile' concept
+     * @return the value of the parameter identified by the key
+     */
+    @Override
+    protected String getValue(String key) {
+        // Start a configuration session if we don't already have one for the key requested
+        // so that we can the initial token. If we already have a session, we can take
+        // the next request token from there.
+        EstablishedSession establishedSession = establishedSessions.getOrDefault(key, null);
+        String sessionToken = establishedSession != null?
+                establishedSession.nextSessionToken :
+                client.startConfigurationSession(StartConfigurationSessionRequest.builder()
+                            .applicationIdentifier(this.application)
+                            .environmentIdentifier(this.environment)
+                            .configurationProfileIdentifier(key)
+                            .build())
+                    .initialConfigurationToken();
+
+        // Get the configuration using the token
+        GetLatestConfigurationResponse response = client.getLatestConfiguration(GetLatestConfigurationRequest.builder()
+                        .configurationToken(sessionToken)
+                .build());
+
+        // Get the next session token we'll use next time we are asked for this key
+        String nextSessionToken = response.nextPollConfigurationToken();
+
+        // Get the value of the key. Note that AppConfig will return null if the value
+        // has not changed since we last asked for it in this session - in this case
+        // we return the value we stashed at last request.
+        String value = response.configuration() != null?
+                response.configuration().asUtf8String() : // if we have a new value, use it
+                    establishedSession != null?
+                            establishedSession.lastConfigurationValue : // if we don't but we have a previous value, use that
+                            null; // otherwise we've got no value
```

Comment:
How about using `if` statement instead of using nested ternary operator? IMHO it might be a bit more readable: 

```
        String value = null;
        if(response.configuration() != null) { // if we have a new value, use it
            value = response.configuration().asUtf8String();
        } else if (establishedSession != null){ // if we don't but we have a previous value, use that
            value = establishedSession.lastConfigurationValue;
        }
```

Reply:
It is a big assignment, but I personally find it easier to reason about this way - you assign the variable once in one place and it's easier to refactor for that - you can't have more code creep in between variable definition and assignment of value that might break something. The alternate would also need another branch too - `establishedSession` is also nullable.